### PR TITLE
Backend for Dev Cart

### PR DESCRIPTION
**Pull request by hntejas**

### CODE REVIEW

File: routes/brands.router.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  brands = await Brand.insertMany(req.body.brands);
+  res.json({
+    success: true,
+    data: {
+      message: "Brands created",
+      brands: brands,
+      count: brands && brands.length
+    }
+  });
+});
```
Comment (prithviBytes) [inline routes/brands.router.js:25]: Hey,
Not sure if it is a best practice, but wouldn't chaining the route with the same path help us structure the code better.
Something like this

```
router.route("/")
.get((req,res) => {
   code
})
.post((req,res) => {
  code
})
```

Or is there a reason so as to not do it?
If chaining is a better approach then could you do it for all the files.

Comment (hntejas) [reply routes/brands.router.js:25]: Thanks for the review. Yes I agree that chaining will provide a better structure. I'll do the changes in all files.

**Review by prithviBytes — CHANGES_REQUESTED** [review]
Hi Tejas,
Very well-written code. Couldn't find any enhancements TBH, Just one.
Thanks,
Prithvi

**Comment by hntejas** [comment]
> Hi Tejas,
> Very well-written code. Couldn't find any enhancements TBH, Just one.
> Thanks,
> Prithvi

Thanks for the review Prithvi, I've updated to code to follow the suggested best practice. Please feel free to check and let know if anything else can be done.

**Review by prithviBytes — APPROVED** [review]
Hey, 
Everything looks good.
Approving the changes from my end!

File: data/products.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
+export const products = [
```
Comment (shubham-ghuge) [inline data/products.js:1]: May be clean this file if not in use.

File: routes/auth.router.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+          message: "User not registered"
+        }
+    })
+  }  
+})
+
+async function checkIfUserExists(email) {
+  const user = await User.find({ email: email });
+  return user[0];
+}
```
Comment (shubham-ghuge) [inline routes/auth.router.js]: instead of find you can use findOne right?

File: routes/auth.router.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    const hashedPassword = bcrypt.hashSync(password, bcryptSalt);   
+    const user = await User.create({ name: name, email: email, password: hashedPassword });
+    const cart = await Cart.create({uid: user._id});
+    const wishlist = await Wishlist.create({uid: user._id});
+    const token = generateJWT(user._id.toString());
+
+    res.json({
+      success: true,
+      token: token
+    });
```
Comment (shubham-ghuge) [inline routes/auth.router.js]: res.json({ success:true, token }) would also work

File: routes/address.router.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  const addresses = await Address.find({uid: req.uid});
+  res.json({
+    success: true,
+    addresses: addresses
+  });
+});
+
+router.post("/", async (req, res) => {
+  try{
+    console.log(req.uid);
```
Comment (shubham-ghuge) [inline routes/address.router.js]: remove console.log

File: transformer.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  const cartObj = cart.toObject();
+
+  cartObj.cartLines = cartObj.cartLines.map(cartLine => ({
+    ...cartLine,
+    _id: undefined,
+    product: {...cartLine.product, id: cartLine.product._id, _id: undefined}
+  }));
+  cartObj._id = undefined;
+  cartObj.uid = undefined;
+  cartObj.__v = undefined;
```
Comment (shubham-ghuge) [inline transformer.js]: this code could be extracted to function to avoid DRY code

**Review by shubham-ghuge — CHANGES_REQUESTED** [review]
hey tejas,
The code and overall logic is really upto the mark, kudos to that!
i have given few Suggestions in comments, also i would highly recommend you to use prettier to format the code,also the transformer.js can be stored in a separate folder such as controllers to keep the logic out from the routers folder.
Thank you,
Shubham .

File: models/brand.model.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+const mongoose = require("mongoose");
+
+const {Schema} = mongoose;
+
+const BrandSchema = new Schema({
+  name: String,
+  imgUrl: String
```
Comment (rohan9896) [inline models/brand.model.js]: Instead of string, you should use mongoose.SchemaTypes.Url

File: models/address.model.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+const mongoose = require("mongoose");
+
+const {Schema} = mongoose;
+
+const AddressSchema = new Schema({
+  uid: {type: Schema.Types.ObjectId, ref: "User"},
+  addressLine1: String,
```
Comment (rohan9896) [inline models/address.model.js]: You can add required validation

File: models/address.model.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  addressLine2: String,
+  city: String,
+  state: String,
+  zip: String,
+  phone: String
```
Comment (rohan9896) [inline models/address.model.js]: phone number should be unique, so you can add unique: true here

File: utils/db.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+const mongoose = require('mongoose');
+const {mongoURL} = require('../data/configs');
+mongoose.plugin(require('meanie-mongoose-to-json'));
+
+module.exports.init = async () => {
+  try{
+    console.log(mongoURL)
```
Comment (rohan9896) [inline utils/db.js]: remove this console.log if not required

File: routes/address.router.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+router.get("/", async (req, res) => {
+  try{
+    const addresses = await Address.find({uid: req.uid});
+    res.json({
+      success: true,
+      addresses: addresses
+    });
+  }catch(e){
+    res.status(500).json({
+      error: {
```
Comment (rohan9896) [inline routes/address.router.js:16]: if you are using success: true in try, then you should also add success: false in res.json of catch block

File: routes/address.router.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+router.delete("/:addressId", async (req, res) => {
+  try{
+    const {addressId} = req.params;
+    await Address.deleteOne({uid: req.uid, _id: addressId})
+    res.json({
+      success: true
+    })
+  }catch(e){
+    res.status(500).json({
+      error: {
```
Comment (rohan9896) [inline routes/address.router.js:70]: Add success: false in this entire file

File: models/category.model.js [hunk 13]
```diff
@@ -0,0 +1,12 @@
+const mongoose = require("mongoose");
+
+const {Schema} = mongoose;
+
+const CategorySchema = new Schema({
+  name: String,
+  imgUrl: String
```
Comment (rohan9896) [inline models/category.model.js]: Add mongoose.SchemaTypes.Url

File: routes/order.router.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+router.post("/", async (req, res) => {
+  try {
+    const cart = await Cart.findOne({ uid: req.uid });
+    const populatedCart = await cart.populate({
+      path: 'cartLines.product'
+    }).execPopulate();
+    const totalPrice = getTotalPrice(populatedCart);
+    const order = await Order.create({ uid: req.uid, products: cart.cartLines, totalPrice: totalPrice });
+    cart.cartLines = [];
+    const updatedCart = await cart.save();
```
Comment (rohan9896) [inline routes/order.router.js]: As you are not sending this updatedCart variable in res.json, so you can directly do await cart.save()

**Review by rohan9896 — CHANGES_REQUESTED** [review]
Hey Tejas, the code looks very clean and consistent, I have suggested some changes mainly in the model files as you haven't done schema validation which is good to have!

File: models/brand.model.js (hunk 7 above)
Comment (hntejas) [reply models/brand.model.js]: mongoose.SchemaTypes.Url does not come out of the box with mongoose but depends on an external library mongoose-type-url which was released 8 months ago. I would like to keep less dependencies on the early stage libraries at this point, hence continuing to use string.

File: models/address.model.js (hunk 8 above)
Comment (hntejas) [reply models/address.model.js]: Added validations

File: models/address.model.js (hunk 9 above)
Comment (hntejas) [reply models/address.model.js]: I'm considering the possibility of same phone number, ex: I would have 2 addresses, one for my home and one for my office. But the contact would still be myself, hence not choosing the phone to be unique

File: utils/db.js (hunk 10 above)
Comment (hntejas) [reply utils/db.js]: removed

File: routes/address.router.js (hunk 11 above)
Comment (hntejas) [reply routes/address.router.js:16]: added the success: false to all the places required in the code

File: routes/address.router.js (hunk 12 above)
Comment (hntejas) [reply routes/address.router.js:70]: added the success: false to all the places required in the code

File: models/category.model.js (hunk 13 above)
Comment (hntejas) [reply models/category.model.js]: using String, reason same as mentioned above

File: routes/order.router.js (hunk 14 above)
Comment (hntejas) [reply routes/order.router.js]: I've made the change as suggested

File: transformer.js (hunk 6 above)
Comment (hntejas) [reply transformer.js]: This has been moved to a function

File: routes/address.router.js (hunk 5 above)
Comment (hntejas) [reply routes/address.router.js]: removed

File: routes/auth.router.js (hunk 4 above)
Comment (hntejas) [reply routes/auth.router.js]: yes, keeping the key value to pair to ease the readability

File: routes/auth.router.js (hunk 3 above)
Comment (hntejas) [reply routes/auth.router.js]: yes findOne makes the right fit, changes have been made

**Comment by hntejas** [comment]
> hey tejas,
> The code and overall logic is really upto the mark, kudos to that!
> i have given few Suggestions in comments, also i would highly recommend you to use prettier to format the code,also the transformer.js can be stored in a separate folder such as controllers to keep the logic out from the routers folder.
> Thank you,
> Shubham .

Hi Shubham,
Thanks for your time in reviewing and pointing out some good suggestions. I've made the changes accordingly and added the comments, please do have a look at it. 
On the transformer.js - this was not having any business logic to be placed as a controller, hence I have placed it in the utils folder

**Comment by hntejas** [comment]
> Hey Tejas, the code looks very clean and consistent, I have suggested some changes mainly in the model files as you haven't done schema validation which is good to have!

Hey Rohan,
Thanks for taking time to review my code and pointing some good suggestions. I've made the changes accordingly and commented on the same. Please do have a look at it.

**Review by shubham-ghuge — APPROVED** [review]
Hey Tejas,
Thank you for considering the changes,
approving from my end!

**Pull request merged**